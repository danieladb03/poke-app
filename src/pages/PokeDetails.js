import styled from "styled-components";
import Layout from "./Layout";
import Container from "../core/Container";
import Header from "../core/Header";
import statsImg from "../assets/Stats.svg";
import {PokeImage} from "../core/PokeCard";
import {Subtitle, BoldSubtitle, TextContent} from "../core/typography";
import {
  useFetchPokemon,
  useFetchPokeSpecies,
} from "../services/poke-api/pokemon";
import {useMemo} from "react"; // para hacer cache de valores
import {Link} from "wouter";
import Loader from "../core/loader";

const categoryImg = "https://cdn.bulbagarden.net/upload/";
const cateogryUrls = {
  normal: "9/95/Normal_icon_SwSh.png",
  fighting: "3/3b/Fighting_icon_SwSh.png",
  flying: "b/b5/Flying_icon_SwSh.png",
  poison: "8/8d/Poison_icon_SwSh.png",
  ground: "2/27/Ground_icon_SwSh.png",
  rock: "1/11/Rock_icon_SwSh.png",
  bug: "9/9c/Bug_icon_SwSh.png",
  ghost: "0/01/Ghost_icon_SwSh.png",
  steel: "0/09/Steel_icon_SwSh.png",
  fire: "a/ab/Fire_icon_SwSh.png",
  water: "8/80/Water_icon_SwSh.png",
  grass: "a/a8/Grass_icon_SwSh.png",
  electric: "7/7b/Electric_icon_SwSh.png",
  psychic: "7/73/Psychic_icon_SwSh.png",
  ice: "1/15/Ice_icon_SwSh.png",
  dragon: "7/70/Dragon_icon_SwSh.png",
  dark: "d/d5/Dark_icon_SwSh.png",
  fairy: "c/c6/Fairy_icon_SwSh.png",
  unknown: "5/56/UnknownIC_PBR.png",
  shadow: "5/56/UnknownIC_PBR.png",
};
const getCategoryUrl = (type) => {
  if (!type) return "";
  return `${categoryImg}${cateogryUrls[type] || cateogryUrls.unknown}`;
};

const Wrapper = styled(Container)`
  width: calc(100% - 298px);
`;

const PokeDetailsWrapper = styled(Container)`
  justify-content: "space-between";
  height: calc(100vh - 65px);
  padding: 55px 10px 0px 10px;
  overflow-y: auto;
`;

const PokeCharacteristics = styled(Container)`
  width: 50%;
  background: url(${(props) => getCategoryUrl(props.type)}) no-repeat center;
  background-size: 70px 70px;
`;

const PokeStats = styled(Container)`
  width: 50%;
  background: url(${statsImg}) no-repeat center;

  ${Container} {
    .StatLabel {
      white-space: nowrap;
      text-overflow: ellipsis;
      overflow: hidden;
      width: 60px;
      text-align: center;
    }
  }
`;

const StatBar = styled.div`
  width: 30px;
  background-color: #f2cd3c;
  margin: 0px 20px 10px 20px;
  height: ${(props) => props.statHeight};
`;

const PokeDetails = ({params: {pokeId}}) => {
  const isLoading = true;
  const {pokemon, isLoading1, isError} = useFetchPokemon(pokeId);
  const {
    pokedescription,
    pokecategory,
    isLoading1: isLoadingDescription, // cambiando nombre de variable
  } = useFetchPokeSpecies(pokeId);
  // TODO mover al fetch
  const description = useMemo(() => {
    return pokedescription?.find((desc) => desc.language.name === "en")
      ?.flavor_text;
  }, [pokedescription]);
  const category = useMemo(() => {
    return pokecategory?.find((cate) => cate.language.name === "en")?.genus;
  }, [pokecategory]);
  console.log("Pokemon", pokemon);
  console.log("description", description);
  console.log("category", category);
  return (
    <Layout>
      <Wrapper flexDirection="column">
        <Header>
          <Container justifyContent="flex-end" alignItems="center">
            <Link href="/">
              <span class="material-icons md-light">arrow_back</span>
            </Link>
          </Container>
        </Header>
        <PokeDetailsWrapper flexWrap="wrap">
          <Container
            style={{width: "50%"}}
            flexDirection="column"
            alignItems="center"
            justifyContent="center"
          >
            <Loader
              isLoading={isLoading}
              width={200}
              height={200}
              style={{marginBottom: "10px"}}
            >
              <PokeImage
                src={pokemon.img}
                alt="Poke img"
                imageHeight="170px"
              ></PokeImage>
            </Loader>
            <Loader
              isLoading={isLoading}
              width={200}
              height={16}
              style={{marginBottom: "10px"}}
            >
              <BoldSubtitle color="#000000" textTransform="capitalize">
                {pokemon.name}
              </BoldSubtitle>
            </Loader>
          </Container>
          <Container
            style={{width: "50%"}}
            flexDirection="column"
            justifyContent="center"
          >
            <Container
              flexDirection="column"
              justifyContent="center"
              margin="0px 60px"
            >
              <Loader
                isLoading={isLoading}
                width={200}
                height={16}
                style={{marginBottom: "10px"}}
              >
                <BoldSubtitle
                  color="#000000"
                  textTransform="capitalize"
                  margin="15px 0px"
                >
                  About {pokemon.name}
                </BoldSubtitle>
              </Loader>
              <Loader
                isLoading={isLoading}
                width={400}
                height={32}
                style={{marginBottom: "10px"}}
              >
                <TextContent color="#000000">{description}</TextContent>
              </Loader>
            </Container>
          </Container>
          <PokeCharacteristics
            alignItems="center"
            flexWrap="wrap"
            justifyContent="center"
            type={
              pokemon?.characteristics?.types &&
              pokemon.characteristics.types[0].name
            }
          >
            <Container
              width="60%"
              height="80%"
              alignItems="center"
              flexWrap="wrap"
            >
              <Container
                style={{width: "50%"}}
                flexDirection="column"
                alignItems="center"
              >
                <Loader
                  isLoading={isLoading}
                  width={70}
                  height={14}
                  style={{marginBottom: "10px"}}
                >
                  <Subtitle color="#787878">Height:</Subtitle>
                </Loader>
                <Loader
                  isLoading={isLoading}
                  width={70}
                  height={14}
                  style={{marginBottom: "10px"}}
                >
                  <Subtitle color="#787878">
                    {pokemon.characteristics.height}"
                  </Subtitle>
                </Loader>
              </Container>
              <Container
                style={{width: "50%"}}
                flexDirection="column"
                alignItems="center"
                justifyContent="center"
              >
                <Loader
                  isLoading={isLoading}
                  width={70}
                  height={14}
                  style={{marginBottom: "10px"}}
                >
                  <Subtitle color="#787878">Category:</Subtitle>
                </Loader>
                <Loader
                  isLoading={isLoading}
                  width={70}
                  height={14}
                  style={{marginBottom: "10px"}}
                >
                  <Subtitle color="#787878">{category}</Subtitle>
                </Loader>
              </Container>
              <Container
                style={{width: "50%"}}
                flexDirection="column"
                alignItems="center"
              >
                <Loader
                  isLoading={isLoading}
                  width={70}
                  height={14}
                  style={{marginBottom: "10px"}}
                >
                  <Subtitle color="#787878">Weight:</Subtitle>
                </Loader>
                <Loader
                  isLoading={isLoading}
                  width={70}
                  height={14}
                  style={{marginBottom: "10px"}}
                >
                  <Subtitle color="#787878">
                    {pokemon.characteristics.weight} lbs
                  </Subtitle>
                </Loader>
              </Container>
              <Container
                style={{width: "50%"}}
                flexDirection="column"
                alignItems="center"
              >
                <Loader
                  isLoading={isLoading}
                  width={70}
                  height={14}
                  style={{marginBottom: "10px"}}
                >
                  <Subtitle color="#787878">Abilities:</Subtitle>
                </Loader>
                <Loader
                  isLoading={isLoading}
                  width={70}
                  height={14}
                  style={{marginBottom: "10px"}}
                >
                  {!isLoading &&
                    pokemon.characteristics.abilities.map((ability, k) => (
                      <Subtitle
                        key={k}
                        color="#787878"
                        textTransform="capitalize"
                      >
                        {ability.ability.name}
                      </Subtitle>
                    ))}
                </Loader>
              </Container>
            </Container>
          </PokeCharacteristics>
          <PokeStats
            flexDirection="column"
            alignItems="center"
            justifyContent="center"
            flexWrap="wrap"
          >
            <Container
              style={{width: "400px", height: "140px"}}
              alignItems="flex-end"
              justifyContent="center"
            >
              <Loader
                isLoading={isLoading}
                width={70}
                height={14}
                style={{marginBottom: "10px"}}
              >
                {!isLoading &&
                  pokemon.stats.map((stat, k) => (
                    <Container
                      key={k}
                      flexDirection="column"
                      alignItems="center"
                    >
                      <StatBar statHeight={`${stat.base_stat}px`}></StatBar>
                      <BoldSubtitle
                        className="StatLabel"
                        color="#787878"
                        textTransform="capitalize"
                      >
                        {stat.stat.name}
                      </BoldSubtitle>
                    </Container>
                  ))}
              </Loader>
            </Container>
          </PokeStats>
        </PokeDetailsWrapper>
      </Wrapper>
    </Layout>
  );
};

export default PokeDetails;
