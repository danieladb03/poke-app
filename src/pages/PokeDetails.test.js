import {render, screen} from "@testing-library/react";
import PokeDetails from "./PokeDetails";
import {
  useFetchPokemon,
  useFetchPokeSpecies,
} from "../services/poke-api/pokemon";
import renderer from "react-test-renderer";

jest.mock("../services/poke-api/pokemon");

describe("PokeDetails", () => {
  const expectedHeight = "height";
  const expectedWeight = "weight";
  const expectedAbility = "ability";
  const expectedStat = "hp";
  const pokeResponse = {
    pokemon: {
      name: "daniela",
      img: "img",
      characteristics: {
        height: expectedHeight,
        abilities: [{ability: {name: expectedAbility}}],
        types: [{type: {name: "type"}}],
        weight: expectedWeight,
      },
      stats: [{base_stat: "38", stat: {name: expectedStat}}],
    },
  };
  const expectedDescription = "blablala";
  const expectedCategory = "category";
  const pokeDescResponse = {
    pokedescription: [
      {flavor_text: expectedDescription, language: {name: "en"}},
    ],
    pokecategory: [{language: {name: "en"}, genus: expectedCategory}],
  };

  test("returns pokemon details", () => {
    const params = {pokeId: 1};
    useFetchPokemon.mockReturnValue(pokeResponse);
    useFetchPokeSpecies.mockReturnValue(pokeDescResponse);
    render(<PokeDetails params={params} />);
    const pokeName = screen.getByText(pokeResponse.pokemon.name);

    const tree = renderer.create(<PokeDetails params={params} />).toJSON();
    expect(tree).toMatchSnapshot();

    expect(pokeName).toBeTruthy();
    expect(screen.getByText(`About ${pokeResponse.pokemon.name}`)).toBeTruthy();
    expect(screen.getByText(expectedDescription)).toBeTruthy();
    expect(screen.getByText(`${expectedHeight}"`)).toBeTruthy();
    expect(screen.getByText(`${expectedCategory}`)).toBeTruthy();
    expect(screen.getByText(`${expectedWeight} lbs`)).toBeTruthy();
    expect(screen.getByText(`${expectedAbility}`)).toBeTruthy();
    expect(screen.getByText(`${expectedStat}`)).toBeTruthy();
  });

  test("no type", () => {
    const params = {pokeId: 1};
    const resp = {
      pokemon: {
        name: "daniela",
        img: "img",
        characteristics: {
          height: expectedHeight,
          abilities: [{ability: {name: expectedAbility}}],
          types: [{type: {}}],
          weight: expectedWeight,
        },
        stats: [{base_stat: "38", stat: {name: expectedStat}}],
      },
    };
    useFetchPokemon.mockReturnValue(resp);
    useFetchPokeSpecies.mockReturnValue(pokeDescResponse);
    render(<PokeDetails params={params} />);

    const tree = renderer.create(<PokeDetails params={params} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
