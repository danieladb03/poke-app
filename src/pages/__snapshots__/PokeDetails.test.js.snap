// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokeDetails no type 1`] = `
<div
  className="sc-bdfBwQ eXdLAz"
>
  <div
    className="sc-bdfBwQ sc-jSgupP ea-DoiS eEPkuY"
  >
    <h1
      className="sc-gsTCUz gveNNn"
      color="#FFFFFF"
    >
      Poke
    </h1>
    <img
      className="sc-iBPRYJ djjWhJ"
      src="profile.png"
    />
    <p
      className="sc-gKsewC cqGOif"
    >
      Daniela DB
    </p>
    <p
      className="sc-dlfnbm kVIrGI"
      color="#FFFFFF"
    >
      Poke Master
    </p>
    <p
      className="sc-dlfnbm hrqAOj"
      color="#FFFFFF"
    >
      Lvl 25
    </p>
  </div>
  <div
    className="sc-bdfBwQ sc-jrAGrp hZVvYV hiYJFV"
  >
    <header
      className="sc-fubCfw ePVVUN"
    >
      <div
        className="sc-bdfBwQ blFYpd"
      >
        <span
          className="material-icons md-light"
          href="/"
          onClick={[Function]}
          to={null}
        >
          arrow_back
        </span>
      </div>
    </header>
    <div
      className="sc-bdfBwQ sc-kEjbxe exVEDH hojsNo"
    >
      <div
        className="sc-bdfBwQ fLYtOx"
        style={
          Object {
            "width": "50%",
          }
        }
      >
        <img
          alt="Poke img"
          className="sc-pFZIQ hLbztS"
          src="img"
        />
        <p
          className="sc-hKgILt dzUNgq"
          color="#000000"
        >
          daniela
        </p>
      </div>
      <div
        className="sc-bdfBwQ blZTAu"
        style={
          Object {
            "width": "50%",
          }
        }
      >
        <div
          className="sc-bdfBwQ fClFsI"
        >
          <p
            className="sc-hKgILt gQPthW"
            color="#000000"
          >
            About 
            daniela
          </p>
          <p
            className="sc-eCssSg iBbOBr"
            color="#000000"
          >
            blablala
          </p>
        </div>
      </div>
      <div
        className="sc-bdfBwQ sc-iqHYGH fxSYUT eUgsuE"
      >
        <div
          className="sc-bdfBwQ cePUqa"
          height="80%"
          width="60%"
        >
          <div
            className="sc-bdfBwQ ea-DoiS"
            style={
              Object {
                "width": "50%",
              }
            }
          >
            <p
              className="sc-dlfnbm bHRgVr"
              color="#787878"
            >
              Height:
            </p>
            <p
              className="sc-dlfnbm bHRgVr"
              color="#787878"
            >
              height
              "
            </p>
          </div>
          <div
            className="sc-bdfBwQ fLYtOx"
            style={
              Object {
                "width": "50%",
              }
            }
          >
            <p
              className="sc-dlfnbm bHRgVr"
              color="#787878"
            >
              Category:
            </p>
            <p
              className="sc-dlfnbm bHRgVr"
              color="#787878"
            >
              category
            </p>
          </div>
          <div
            className="sc-bdfBwQ ea-DoiS"
            style={
              Object {
                "width": "50%",
              }
            }
          >
            <p
              className="sc-dlfnbm bHRgVr"
              color="#787878"
            >
              Weight:
            </p>
            <p
              className="sc-dlfnbm bHRgVr"
              color="#787878"
            >
              weight
               lbs
            </p>
          </div>
          <div
            className="sc-bdfBwQ ea-DoiS"
            style={
              Object {
                "width": "50%",
              }
            }
          >
            <p
              className="sc-dlfnbm bHRgVr"
              color="#787878"
            >
              Abilities:
            </p>
            <p
              className="sc-dlfnbm bTYniz"
              color="#787878"
            >
              ability
            </p>
          </div>
        </div>
      </div>
      <div
        className="sc-bdfBwQ sc-crrsfI jeIypN bJdpfI"
      >
        <div
          className="sc-bdfBwQ hdkJVZ"
          style={
            Object {
              "height": "140px",
              "width": "400px",
            }
          }
        >
          <div
            className="sc-bdfBwQ ea-DoiS"
          >
            <div
              className="sc-dQppl dPqAHA"
            />
            <p
              className="sc-hKgILt dgwuDt StatLabel"
              color="#787878"
            >
              hp
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`PokeDetails returns pokemon details 1`] = `
<div
  className="sc-bdfBwQ eXdLAz"
>
  <div
    className="sc-bdfBwQ sc-jSgupP ea-DoiS eEPkuY"
  >
    <h1
      className="sc-gsTCUz gveNNn"
      color="#FFFFFF"
    >
      Poke
    </h1>
    <img
      className="sc-iBPRYJ djjWhJ"
      src="profile.png"
    />
    <p
      className="sc-gKsewC cqGOif"
    >
      Daniela DB
    </p>
    <p
      className="sc-dlfnbm kVIrGI"
      color="#FFFFFF"
    >
      Poke Master
    </p>
    <p
      className="sc-dlfnbm hrqAOj"
      color="#FFFFFF"
    >
      Lvl 25
    </p>
  </div>
  <div
    className="sc-bdfBwQ sc-jrAGrp hZVvYV hiYJFV"
  >
    <header
      className="sc-fubCfw ePVVUN"
    >
      <div
        className="sc-bdfBwQ blFYpd"
      >
        <span
          className="material-icons md-light"
          href="/"
          onClick={[Function]}
          to={null}
        >
          arrow_back
        </span>
      </div>
    </header>
    <div
      className="sc-bdfBwQ sc-kEjbxe exVEDH hojsNo"
    >
      <div
        className="sc-bdfBwQ fLYtOx"
        style={
          Object {
            "width": "50%",
          }
        }
      >
        <img
          alt="Poke img"
          className="sc-pFZIQ hLbztS"
          src="img"
        />
        <p
          className="sc-hKgILt dzUNgq"
          color="#000000"
        >
          daniela
        </p>
      </div>
      <div
        className="sc-bdfBwQ blZTAu"
        style={
          Object {
            "width": "50%",
          }
        }
      >
        <div
          className="sc-bdfBwQ fClFsI"
        >
          <p
            className="sc-hKgILt gQPthW"
            color="#000000"
          >
            About 
            daniela
          </p>
          <p
            className="sc-eCssSg iBbOBr"
            color="#000000"
          >
            blablala
          </p>
        </div>
      </div>
      <div
        className="sc-bdfBwQ sc-iqHYGH fxSYUT cfXoQn"
        type="type"
      >
        <div
          className="sc-bdfBwQ cePUqa"
          height="80%"
          width="60%"
        >
          <div
            className="sc-bdfBwQ ea-DoiS"
            style={
              Object {
                "width": "50%",
              }
            }
          >
            <p
              className="sc-dlfnbm bHRgVr"
              color="#787878"
            >
              Height:
            </p>
            <p
              className="sc-dlfnbm bHRgVr"
              color="#787878"
            >
              height
              "
            </p>
          </div>
          <div
            className="sc-bdfBwQ fLYtOx"
            style={
              Object {
                "width": "50%",
              }
            }
          >
            <p
              className="sc-dlfnbm bHRgVr"
              color="#787878"
            >
              Category:
            </p>
            <p
              className="sc-dlfnbm bHRgVr"
              color="#787878"
            >
              category
            </p>
          </div>
          <div
            className="sc-bdfBwQ ea-DoiS"
            style={
              Object {
                "width": "50%",
              }
            }
          >
            <p
              className="sc-dlfnbm bHRgVr"
              color="#787878"
            >
              Weight:
            </p>
            <p
              className="sc-dlfnbm bHRgVr"
              color="#787878"
            >
              weight
               lbs
            </p>
          </div>
          <div
            className="sc-bdfBwQ ea-DoiS"
            style={
              Object {
                "width": "50%",
              }
            }
          >
            <p
              className="sc-dlfnbm bHRgVr"
              color="#787878"
            >
              Abilities:
            </p>
            <p
              className="sc-dlfnbm bTYniz"
              color="#787878"
            >
              ability
            </p>
          </div>
        </div>
      </div>
      <div
        className="sc-bdfBwQ sc-crrsfI jeIypN bJdpfI"
      >
        <div
          className="sc-bdfBwQ hdkJVZ"
          style={
            Object {
              "height": "140px",
              "width": "400px",
            }
          }
        >
          <div
            className="sc-bdfBwQ ea-DoiS"
          >
            <div
              className="sc-dQppl dPqAHA"
            />
            <p
              className="sc-hKgILt dgwuDt StatLabel"
              color="#787878"
            >
              hp
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
